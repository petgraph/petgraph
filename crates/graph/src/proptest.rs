use alloc::sync::Arc;
use core::fmt::Debug;

use petgraph_core::deprecated::{edge::EdgeType, index::IndexType};
use petgraph_proptest::default::graph_strategy;
use proptest::{
    arbitrary::Arbitrary, collection::SizeRange, prelude::BoxedStrategy, strategy::Strategy,
};

use crate::Graph;

/// `Arbitrary` for `Graph` creates a graph by selecting a node count
///
/// The result will be simple graph or digraph, self loops
/// possible, no parallel edges.
///
/// The exact properties of the produced graph is subject to change.
///
/// Requires crate feature `"proptest"`
impl<N, E, Ty, Ix> Arbitrary for Graph<N, E, Ty, Ix>
where
    N: Arbitrary + Debug + Clone + 'static,
    E: Arbitrary + Debug + 'static,
    Ty: EdgeType + Send + 'static,
    Ix: IndexType + Send,
{
    type Parameters = ();
    // impl Strategy<Value = Self> is nightly, and therefore not usable here.
    // TODO: revisit once impl_trait_in_assoc_type is stable. (https://github.com/rust-lang/rust/issues/63063)
    type Strategy = BoxedStrategy<Self>;

    // `Graph` has a restriction, that the last node index cannot be used.
    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
        graph_strategy(
            true,
            false,
            0..Ix::MAX.as_usize(),
            Some(Arc::new(|max| {
                SizeRange::new(0..=usize::min(max.pow(2), Ix::MAX.as_usize() - 1))
            })),
        )
        .boxed()
    }
}
