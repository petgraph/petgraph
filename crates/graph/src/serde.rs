//! Serde support for `Graph` and `StableGraph`
//!
//! ## Serialization Format
//!
//! The serialization format is as follows, in pseudocode:
//!
//! ```text
//! Graph {
//!     nodes: [N],
//!     node_holes: [NodeIndex<Ix>],
//!     edge_property: EdgeProperty,
//!     edges: [Option<(NodeIndex<Ix>, NodeIndex<Ix>, E)>]
//! }
//! ```
//!
//! Node indices are serialized as integers and are fixed size for binary formats, so the Ix
//! parameter matters there.
//!
//! A stable graph serialization that obeys these restrictions (effectively, it has no interior
//! vacancies) can de deserialized as a [`Graph`].
//!
//! ## [`Graph`] Restrictions
//!
//! `node_holes` is always empty and `edges` are always `Some`.

use alloc::{string::ToString, vec::Vec};
use core::{
    fmt::{Display, Formatter},
    marker::PhantomData,
};

use funty::Integral;
use petgraph_core::{edge::EdgeType, index::IndexType};
use serde::{
    de::Error as _,
    ser::{SerializeStruct, SerializeTuple},
    Deserialize, Deserializer, Serialize, Serializer,
};

use super::{EdgeIndex, NodeIndex};
use crate::{Edge, Graph, Node};

#[derive(Debug)]
pub enum InvalidError {
    Node {
        index: usize,
        length: usize,
    },
    Hole {
        index: usize,
    },
    Length {
        type_: &'static str,
        length: usize,
        max: usize,
    },
    Direction {
        expected: EdgeProperty,
        received: EdgeProperty,
    },
}

impl Display for InvalidError {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Node { index, length } => f.write_fmt(format_args!(
                "invalid value: node index `{index}` does not exist in graph with length \
                 `{length}`",
            )),
            Self::Hole { index } => f.write_fmt(format_args!(
                "invalid value: node hole `{index}` is not allowed.",
            )),
            Self::Length { type_, length, max } => f.write_fmt(format_args!(
                "invalid value: {type_} length `{length}` exceeds maximum of `{max}`",
            )),
            Self::Direction { expected, received } => f.write_fmt(format_args!(
                "invalid value: expected {expected} graph, but received {received} graph",
            )),
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EdgeProperty {
    Undirected,
    Directed,
}

impl Display for EdgeProperty {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Undirected => f.write_str("undirected"),
            Self::Directed => f.write_str("directed"),
        }
    }
}

impl EdgeProperty {
    pub(crate) fn from_type<T>() -> Self
    where
        T: EdgeType,
    {
        if T::is_directed() {
            Self::Directed
        } else {
            Self::Undirected
        }
    }
}

impl<Ix> Serialize for NodeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<Ix> Serialize for EdgeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<E, Ix> Serialize for Edge<E, Ix>
where
    E: Serialize,
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut tuple = serializer.serialize_tuple(3)?;

        tuple.serialize_element(&self.source())?;
        tuple.serialize_element(&self.target())?;
        tuple.serialize_element(&self.weight)?;

        tuple.end()
    }
}

impl<N, Ix> Serialize for Node<N, Ix>
where
    N: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.weight.serialize(serializer)
    }
}

impl<N, E, Ty, Ix> Serialize for Graph<N, E, Ty, Ix>
where
    N: Serialize,
    E: Serialize,
    Ty: EdgeType,
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut struct_ = serializer.serialize_struct("Graph", 4)?;

        struct_.serialize_field("nodes", &self.nodes)?;
        struct_.serialize_field("node_holes", &[(); 0])?;
        struct_.serialize_field("edge_property", &EdgeProperty::from_type::<Ty>())?;
        struct_.serialize_field("edges", &self.edges)?;

        struct_.end()
    }
}

impl<'de, Ix> Deserialize<'de> for NodeIndex<Ix>
where
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ix::deserialize(deserializer).map(NodeIndex)
    }
}

impl<'de, Ix> Deserialize<'de> for EdgeIndex<Ix>
where
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ix::deserialize(deserializer).map(EdgeIndex)
    }
}

#[derive(Deserialize)]
#[serde(rename = "Edge")]
struct EdgeRemote<E, Ix: IndexType>(NodeIndex<Ix>, NodeIndex<Ix>, E);

impl<'de, E, Ix> Deserialize<'de> for Edge<E, Ix>
where
    E: Deserialize<'de>,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let EdgeRemote(source, target, weight) = EdgeRemote::deserialize(deserializer)?;

        Ok(Edge {
            weight,
            node: [source, target],
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Node")]
struct NodeRemote<N>(N);

impl<'de, N, Ix> Deserialize<'de> for Node<N, Ix>
where
    N: Deserialize<'de>,
    Ix: IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let NodeRemote(weight) = NodeRemote::deserialize(deserializer)?;

        Ok(Node {
            weight,
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Graph")]
struct RemoteGraph<N, E, Ix: IndexType> {
    nodes: Vec<Node<N, Ix>>,
    // always empty per serialization format
    #[serde(default)]
    node_holes: [(); 0],
    edge_property: EdgeProperty,
    edges: Vec<Edge<E, Ix>>,
}

impl<'de, N, E, Ty, Ix> Deserialize<'de> for Graph<N, E, Ty, Ix>
where
    N: Deserialize<'de>,
    E: Deserialize<'de>,
    Ty: EdgeType,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let RemoteGraph {
            nodes,
            edge_property,
            edges,
            ..
        } = RemoteGraph::deserialize(deserializer)?;

        if nodes.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "node",
                length: nodes.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        if edges.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "edge",
                length: edges.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let expected = EdgeProperty::from_type::<Ty>();
        if edge_property != expected {
            return Err(InvalidError::Direction {
                expected,
                received: edge_property,
            })
            .map_err(D::Error::custom);
        }

        let mut this = Self {
            nodes,
            edges,
            ty: PhantomData,
        };

        let node_count = this.node_count();
        this.link_edges()
            .map_err(|i| InvalidError::Node {
                index: i.index(),
                length: node_count,
            })
            .map_err(D::Error::custom)?;

        Ok(this)
    }
}
