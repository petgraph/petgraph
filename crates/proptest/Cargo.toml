[package]
name = "petgraph-proptest"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
petgraph-core = { workspace = true, features = ['alloc'] }
proptest = { workspace = true, features = ['alloc'] }
itertools = "0.10.5" # remove if possible

[dev-dependencies]
petgraph-graph = { workspace = true, features = ['proptest'] }
