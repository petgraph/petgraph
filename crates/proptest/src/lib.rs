//! Generic proptest strategies and support code for petgraph graphs
//!
//! Currently only supports the `default` strategy, which is used by the default implementation of
//! all `Arbitrary` impls.
//!
//! In the future more strategies will be added.
// #![no_std]

extern crate alloc;

pub mod binary;
pub mod dag;
pub mod default;
pub mod tournament;
pub mod vtable;
