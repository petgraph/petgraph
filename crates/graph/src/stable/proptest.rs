use core::fmt::Debug;
use std::{collections::BTreeSet, sync::Arc};

use petgraph_core::{edge::EdgeType, index::IndexType};
use petgraph_proptest::default::graph_strategy;
use proptest::{
    arbitrary::Arbitrary,
    bits::BitSetLike,
    collection::{btree_set, SizeRange},
    prelude::{BoxedStrategy, Just, Strategy},
    strategy::{LazyJust, TupleUnion},
};

use crate::{stable::StableGraph, NodeIndex};

/// `Arbitrary` for `Graph` creates a graph by selecting a node count
///
/// The result will be simple graph or digraph, self loops
/// possible, no parallel edges.
///
/// The exact properties of the produced graph is subject to change.
///
/// Requires crate feature `"proptest"`
impl<N, E, Ty, Ix> Arbitrary for StableGraph<N, E, Ty, Ix>
where
    N: Arbitrary + Debug + Clone + 'static,
    E: Arbitrary + Debug + Clone + 'static,
    Ty: EdgeType + Send + 'static,
    Ix: IndexType + Send,
{
    type Parameters = ();
    // impl Strategy<Value = Self> is nightly, and therefore not usable here.
    // TODO: revisit once impl_trait_in_assoc_type is stable. (https://github.com/rust-lang/rust/issues/63063)
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
        // `MAX` is not a valid node index.
        graph_strategy(
            true,
            false,
            0..Ix::MAX.as_usize(),
            Some(Arc::new(|max| {
                SizeRange::new(0..=usize::min(max.pow(2), Ix::MAX.as_usize() - 1))
            })),
        )
        .prop_flat_map(|graph: Self| {
            let nodes = graph.node_count();
            let is_empty = nodes == 0;

            let removed_nodes = TupleUnion::new((
                (u32::from(is_empty), Arc::new(LazyJust::new(BTreeSet::new))),
                (
                    u32::from(!is_empty),
                    Arc::new(btree_set(0..nodes.len(), 0..nodes.len())),
                ),
            ));

            // select a batch of random indices to remove (unique)
            (Just(graph), removed_nodes)
        })
        .prop_map(|(mut graph, remove)| {
            for index in remove {
                graph.remove_node(NodeIndex(Ix::from_usize(index)));
            }

            graph
        })
        .boxed()
    }
}
