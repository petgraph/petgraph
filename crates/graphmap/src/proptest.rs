use alloc::sync::Arc;
use core::fmt::Debug;

use petgraph_core::deprecated::edge::EdgeType;
use petgraph_proptest::{default::graph_strategy_from_vtable, vtable::VTable};
use proptest::{
    arbitrary::Arbitrary, collection::SizeRange, prelude::BoxedStrategy, strategy::Strategy,
};

use crate::{GraphMap, NodeTrait};

fn add_edge_no_return<N, E, Ty>(graph: &mut GraphMap<N, E, Ty>, a: N, b: N, weight: E)
where
    N: NodeTrait,
    Ty: EdgeType,
{
    graph.add_edge(a, b, weight);
}

// GraphMap does not implement `Create` or `Build`, so we need to use the vtable.
fn create_vtable<N, E, Ty>() -> VTable<GraphMap<N, E, Ty>, N, N, E>
where
    N: NodeTrait,
    Ty: EdgeType,
{
    VTable {
        with_capacity: GraphMap::with_capacity,
        add_node: GraphMap::add_node,
        add_edge: add_edge_no_return::<N, E, Ty>,
    }
}

impl<N, E, Ty> Arbitrary for GraphMap<N, E, Ty>
where
    N: NodeTrait + Arbitrary + Clone + Debug + 'static,
    E: Arbitrary + Debug + 'static,
    Ty: EdgeType + 'static,
{
    type Parameters = ();
    // impl Strategy<Value = Self> is nightly, and therefore not usable here.
    // TODO: revisit once impl_trait_in_assoc_type is stable. (https://github.com/rust-lang/rust/issues/63063)
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
        // creating a graphmap with usize::MAX is a bit excessive, so we use u8::MAX instead.
        graph_strategy_from_vtable(
            create_vtable::<N, E, Ty>(),
            true,
            false,
            0..=(u8::MAX as usize),
            Some(Arc::new(|max| {
                SizeRange::new(0..=usize::min(max.pow(2), u8::MAX as usize))
            })),
        )
        .boxed()
    }
}
