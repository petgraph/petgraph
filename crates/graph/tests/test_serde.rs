#![cfg(feature = "serde")]

use std::{collections::BTreeSet, fmt::Debug};

use insta::assert_json_snapshot;
use petgraph_core::{
    edge::{Directed, Direction, EdgeType, Undirected},
    index::{DefaultIx, IndexType},
    visit::{EdgeRef, IntoEdgeReferences, IntoNeighborsDirected, NodeIndexable},
};
use petgraph_graph::{
    stable::{StableDiGraph, StableGraph},
    DiGraph, EdgeIndex, Graph, NodeIndex,
};
#[cfg(feature = "proptest")]
use proptest::prelude::*;
use serde::{de::DeserializeOwned, Deserialize};
use serde_json::json;

fn assert_iter_eq<I, J>(iter1: I, iter2: J)
where
    I: IntoIterator,
    J: IntoIterator,
    I::Item: PartialEq<J::Item> + Debug,
    J::Item: Debug,
{
    let iter1 = iter1.into_iter();
    let iter2 = iter2.into_iter();

    for (item1, item2) in iter1.zip(iter2) {
        assert_eq!(item1, item2);
    }
}

/// graphs are the equal, down to graph indices
/// this is a strict notion of graph equivalence:
///
/// * Requires equal node and edge indices, equal weights
/// * Does not require: edge for node order
fn assert_graph_eq<N, N2, E, Ty, Ix>(graph1: &Graph<N, E, Ty, Ix>, graph2: &Graph<N2, E, Ty, Ix>)
where
    N: PartialEq<N2> + Debug,
    N2: PartialEq<N2> + Debug,
    E: PartialEq + Debug,
    Ty: EdgeType,
    Ix: IndexType,
{
    assert_eq!(graph1.node_count(), graph2.node_count());
    assert_eq!(graph1.edge_count(), graph2.edge_count());

    // same node weights
    assert_iter_eq(
        graph1.raw_nodes().iter().map(|node| &node.weight),
        graph2.raw_nodes().iter().map(|node| &node.weight),
    );

    // same edge weights
    assert_iter_eq(
        graph1.raw_edges().iter().map(|edge| &edge.weight),
        graph2.raw_edges().iter().map(|edge| &edge.weight),
    );

    for edge in graph1.edge_references() {
        let (source, target) = graph2.edge_endpoints(edge.id()).expect("edge not found");

        assert_eq!(edge.source(), source);
        assert_eq!(edge.target(), target);
    }

    for index in graph1.node_indices() {
        let outgoing1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        let outgoing2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Outgoing)
            .collect();

        assert_eq!(outgoing1, outgoing2);

        let incoming1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Incoming)
            .collect();
        let incoming2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Incoming)
            .collect();

        assert_eq!(incoming1, incoming2);
    }
}

/// graphs are the equal, down to graph indices
/// this is a strict notion of graph equivalence:
//
/// * Requires equal node and edge indices, equal weights
/// * Does not require: edge for node order
///
/// # Panics
///
/// Panics if the graphs are not equal.
#[cfg(feature = "stable")]
pub fn assert_stable_graph_eq<N, E, Ix>(
    graph1: &StableGraph<N, E, Directed, Ix>,
    graph2: &StableGraph<N, E, Directed, Ix>,
) where
    N: PartialEq + Debug,
    E: PartialEq + Debug,
    Ix: IndexType,
{
    assert_eq!(graph1.node_count(), graph2.node_count());
    assert_eq!(graph1.edge_count(), graph2.edge_count());

    // same node weights
    assert_iter_eq(
        (0..graph1.node_bound()).map(|i| graph1.node_weight(NodeIndex::new(i))),
        (0..graph2.node_bound()).map(|i| graph2.node_weight(NodeIndex::new(i))),
    );

    let last_edge_g = graph1.edge_references().next_back();
    let last_edge_h = graph2.edge_references().next_back();

    assert_eq!(last_edge_g.is_some(), last_edge_h.is_some());

    if let (Some(edge_graph1), Some(edge_graph2)) = (last_edge_g, last_edge_h) {
        let last_edge_graph1_index = edge_graph1.id().index();
        let last_edge_graph2_index = edge_graph2.id().index();

        // same edge weigths
        assert_iter_eq(
            (0..last_edge_graph1_index).map(|i| graph1.edge_weight(EdgeIndex::new(i))),
            (0..last_edge_graph2_index).map(|i| graph2.edge_weight(EdgeIndex::new(i))),
        );
    }

    for edge_graph1 in graph1.edge_references() {
        let (source, target) = graph2.edge_endpoints(edge_graph1.id()).unwrap();
        assert_eq!(edge_graph1.source(), source);
        assert_eq!(edge_graph1.target(), target);
    }

    for index in graph1.node_indices() {
        let outgoing1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        let outgoing2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        assert_eq!(outgoing1, outgoing2);

        let incoming1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        let incoming2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        assert_eq!(incoming1, incoming2);
    }
}

#[allow(clippy::many_single_char_names)]
fn example<Ty, Ix>() -> Graph<&'static str, i32, Ty, Ix>
where
    Ty: EdgeType,
    Ix: IndexType,
{
    let mut graph = Graph::default();

    let a = graph.add_node("A");
    let b = graph.add_node("B");
    let c = graph.add_node("C");
    let d = graph.add_node("D");
    let e = graph.add_node("E");
    let f = graph.add_node("F");

    graph.extend_with_edges([
        (a, b, 7),
        (c, a, 9),
        (a, d, 14),
        (b, c, 10),
        (d, c, 2),
        (d, e, 9),
        (b, f, 15),
        (c, f, 11),
        (e, f, 6),
    ]);

    // we remove `d` to ensure that holes are handled correctly
    graph.remove_node(d);

    graph
}

#[allow(clippy::many_single_char_names)]
#[cfg(feature = "stable")]
fn example_stable<Ty, Ix>() -> StableGraph<&'static str, i32, Ty, Ix>
where
    Ty: EdgeType,
    Ix: IndexType,
{
    let mut graph = StableGraph::default();

    let a = graph.add_node("A");
    let b = graph.add_node("B");
    let c = graph.add_node("C");
    let d = graph.add_node("D");
    let e = graph.add_node("E");
    let f = graph.add_node("F");

    graph.extend_with_edges([
        (a, b, 7),
        (c, a, 9),
        (a, d, 14),
        (b, c, 10),
        (d, c, 2),
        (d, e, 9),
        (b, f, 15),
        (c, f, 11),
        (e, f, 6),
    ]);

    // we remove `d` to ensure that holes are handled correctly
    graph.remove_node(d);

    graph
}

#[test]
#[cfg(not(miri))]
fn node_str_edges_i32_serialize() {
    let graph: DiGraph<_, _> = example();

    assert_json_snapshot!(&graph);
}

#[test]
fn node_str_edges_i32_deserialize() {
    let graph: DiGraph<_, _> = example();

    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: DiGraph<String, i32> = DiGraph::deserialize(value).expect("failed to deserialize");

    assert_graph_eq(&graph, &graph2);
}

#[test]
#[cfg(not(miri))]
fn node_null_edges_null_serialize() {
    let graph: DiGraph<(), ()> = example().map(|_, _| (), |_, _| ());

    assert_json_snapshot!(&graph);
}

#[test]
fn node_null_edges_null_deserialize() {
    let graph: DiGraph<(), ()> = example().map(|_, _| (), |_, _| ());
    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: DiGraph<(), ()> = DiGraph::deserialize(value).expect("failed to deserialize");

    assert_graph_eq(&graph, &graph2);
}

#[test]
#[cfg(feature = "stable")]
#[cfg(not(miri))]
fn stable_node_str_edges_i32_serialize() {
    let graph: StableDiGraph<_, _> = example_stable();

    assert_json_snapshot!(&graph);
}

#[test]
#[cfg(feature = "stable")]
fn stable_node_str_edges_i32_deserialize() {
    let graph: StableDiGraph<_, _> =
        example_stable().map(|_, weight| (*weight).to_owned(), |_, weight| *weight);

    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: StableDiGraph<String, i32> =
        StableDiGraph::deserialize(value).expect("failed to deserialize");

    assert_stable_graph_eq(&graph, &graph2);
}

#[test]
#[cfg(feature = "stable")]
#[cfg(not(miri))]
fn stable_node_null_edges_null_serialize() {
    let graph: StableDiGraph<(), ()> = example_stable().map(|_, _| (), |_, _| ());

    assert_json_snapshot!(&graph);
}

#[test]
#[cfg(feature = "stable")]
fn stable_node_null_edges_null_deserialize() {
    let graph: StableDiGraph<(), ()> = example_stable().map(|_, _| (), |_, _| ());
    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: StableDiGraph<(), ()> =
        StableDiGraph::deserialize(value).expect("failed to deserialize");

    assert_stable_graph_eq(&graph, &graph2);
}

#[derive(Debug, Copy, Clone)]
enum Expected {
    Ok,
    Error { error: &'static str },
}

fn assert_inputs<N, E, Ty, Ix>(expected: &[Expected])
where
    N: DeserializeOwned + Debug,
    E: DeserializeOwned + Debug,
    Ty: EdgeType,
    Ix: IndexType + DeserializeOwned,
{
    insta::glob!("inputs/serde/*.json", |path| {
        let content = std::fs::read_to_string(path).expect("failed to read file");
        let index = path
            .file_stem()
            .expect("failed to get file stem")
            .to_string_lossy()
            .split('-')
            .next()
            .expect("failed to get file stem")
            .parse::<usize>()
            .expect("failed to parse file stem");

        let expected = expected[index];

        let graph = serde_json::from_str::<Graph<N, E, Ty, Ix>>(&content);

        match expected {
            Expected::Ok => {
                graph.expect("failed to deserialize");
            }
            Expected::Error {
                error: expected, ..
            } => {
                let error = graph.expect_err("expected error");
                assert_eq!(error.to_string(), *expected);
            }
        }
    });
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_default_null() {
    assert_inputs::<(), (), Directed, DefaultIx>(&[
        Expected::Ok,
        Expected::Error {
            error: "invalid value: node index `5` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: "invalid value: node index `300` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_default_null_undirected() {
    assert_inputs::<(), (), Undirected, DefaultIx>(&[
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_u8_null() {
    assert_inputs::<(), (), Directed, u8>(&[
        Expected::Ok,
        Expected::Error {
            error: "invalid value: node index `5` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: "invalid value: integer `300`, expected u8 at line 18 column 9",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_default_str_i32() {
    assert_inputs::<String, i32, Directed, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Ok,
    ]);
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_default_str_i32_undirected() {
    assert_inputs::<String, i32, Undirected, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
    ]);
}

#[test]
#[cfg(not(miri))]
fn snapshot_deserialize_u8_str_i32() {
    assert_inputs::<String, u8, Directed, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Ok,
    ]);
}

#[test]
fn error_on_too_many_nodes() {
    // The last index (0xFF) is reserved, and cannot be used as a node index.
    let max = u8::MAX;
    let nodes = (0..=max).collect::<Vec<_>>();

    let value = json!({
        "nodes": nodes,
        "edge_property": "directed",
        "edges": []
    });

    let error =
        DiGraph::<u64, (), u8>::deserialize(value).expect_err("expected deserialization to fail");

    assert_eq!(
        error.to_string(),
        "invalid value: node length `256` exceeds maximum of `255`"
    );
}

#[test]
fn error_on_too_many_edges() {
    // The last index (0xFF) is reserved, and cannot be used as an edge index.
    let max = u8::MAX;
    let edges = (0..=max).map(|_| (0, 1, ())).collect::<Vec<_>>();

    let value = json!({
        "nodes": [0, 1],
        "edge_property": "directed",
        "edges": edges
    });

    let error =
        DiGraph::<u64, (), u8>::deserialize(value).expect_err("expected deserialization to fail");

    assert_eq!(
        error.to_string(),
        "invalid value: edge length `256` exceeds maximum of `255`"
    );
}

#[cfg(feature = "proptest")]
#[cfg(not(miri))]
proptest! {
    #[test]
    fn roundtrip(graph in any::<Graph<i8, i8, Directed, u8>>()) {
        let value = serde_value::to_value(graph.clone()).expect("failed to serialize");
        let deserialized = Graph::<i8, i8, Directed, u8>::deserialize(value).expect("failed to deserialize");

        assert_graph_eq(&graph, &deserialized);
    }

    #[test]
    #[cfg(feature = "stable")]
    fn roundtrip_stable_graph(graph in any::<StableGraph<i8, i8, Directed, u8>>()) {
        let value = serde_value::to_value(graph.clone()).expect("failed to serialize");
        let deserialized = StableGraph::<i8, i8, Directed, u8>::deserialize(value).expect("failed to deserialize");

        assert_stable_graph_eq(&graph, &deserialized);
    }

    #[test]
    fn roundtrip_enlarge(graph in any::<Graph<i8, i8, Directed, u8>>()) {
        let value = serde_value::to_value(graph.clone()).expect("failed to serialize");
        let deserialized = Graph::<i8, i8, Directed, u16>::deserialize(value).expect("failed to deserialize");

        let value = serde_value::to_value(deserialized).expect("failed to serialize");
        let reserialized = Graph::<i8, i8, Directed, u8>::deserialize(value).expect("failed to deserialize");

        assert_graph_eq(&graph, &reserialized);
    }

    // We cannot do graph -> stable-graph -> graph, as it is format dependent.
    // Some formats encode `Option<T>` differently to `T`/`null`. (Example: serde-value).
    // While possible for `serde-json` (as `T` / `null` are equivalent to `Option<T>`),
    // it is not for e.g. `bincode`.
    // Due to this fact no property-based test is used.
}
