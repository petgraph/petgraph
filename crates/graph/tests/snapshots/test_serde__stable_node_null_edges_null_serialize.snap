---
source: crates/graph/tests/test_serde.rs
expression: "&graph"
---
{
  "nodes": [
    null,
    null,
    null,
    null,
    null
  ],
  "node_holes": [
    3
  ],
  "edge_property": "directed",
  "edges": [
    [
      0,
      1,
      null
    ],
    [
      2,
      0,
      null
    ],
    null,
    [
      1,
      2,
      null
    ],
    null,
    null,
    [
      1,
      5,
      null
    ],
    [
      2,
      5,
      null
    ],
    [
      4,
      5,
      null
    ]
  ]
}
